File: bitlap-core/src/test/kotlin/org/bitlap/core/test/base/SqlChecker.kt
/* Copyright (c) 2023 bitlap.org */
package org.bitlap.core.test.base

import io.kotest.assertions.Actual
import io.kotest.assertions.Expected
import io.kotest.assertions.failure
import io.kotest.assertions.print.print
import io.kotest.matchers.shouldBe
import org.bitlap.common.utils.SqlEx.toTable
import org.bitlap.common.utils.internal.DBTable
import org.bitlap.core.catalog.metadata.Database
import org.bitlap.core.sql.QueryExecution

/**
 * Some sql check utils
 */
interface SqlChecker {

    /**
     * execute sql statement
     */
    fun sql(statement: String, currentSchema: String = Database.DEFAULT_DATABASE): SqlResult {
        val rs = QueryExecution(statement, currentSchema).execute()
        return SqlResult(statement, rs.data.toTable())
    }

    /**
     * check rows
     */
    fun checkRows(statement: String, rows: List<List<Any?>>, currentSchema: String = Database.DEFAULT_DATABASE) {
        val result = sql(statement, currentSchema).result
        try {
            result.size shouldBe rows.size
            result.zip(rows).forEach { (r1, r2) ->
                r1.size shouldBe r2.size
                r1.zip(r2).forEach { (v1, v2) ->
                    v1 shouldBe v2
                }
            }
        } catch (e: Throwable) {
            when (e) {
                is AssertionError ->
                    throw failure(Expected(rows.print()), Actual(result.print()))
                else ->
                    throw e
            }
        }
    }
}

class SqlResult(private val statement: String, private val table: DBTable) {

    internal val result: List<List<Any?>> by lazy {
        mutableListOf<List<Any?>>().apply {
            (0 until table.rowCount).forEach { r ->
                add(table.columns.map { it.typeValues[r] })
            }
        }
    }

    val size = this.result.size

    fun show(): SqlResult {
        table.show()
        return this
    }

    override fun toString(): String {
        return this.result.toString()
    }

    override fun hashCode(): Int {
        return this.result.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return this.result == other
    }
}

File: bitlap-core/src/test/scala/org/bitlap/core/test/base/SqlChecker.scala
/*
 * Copyright 2020-2023 IceMimosa, jxnu-liguobin and the Bitlap Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.bitlap.core.test.base

import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters._

import org.bitlap.common.utils.SqlEx
import org.bitlap.common.utils.SqlEx.toTable
import org.bitlap.common.utils.internal.DBTable
import org.bitlap.core.catalog.metadata.Database
import org.bitlap.core.sql.QueryExecution

import org.scalatest.matchers.should.Matchers._

/** Some sql check utils
 */
trait SqlChecker {

  /** execute sql statement
   */
  def sql(statement: String, session: SqlSession = SqlSession()): SqlResult = {
    val rs = QueryExecution(statement, session.currentSchema).execute()
    session.currentSchema = rs.currentSchema
    SqlResult(statement, SqlEx.toTable(rs.data))
  }

  /** check rows
   */
  def checkRows(statement: String, rows: List[List[Any]], session: SqlSession = SqlSession()): Unit = {
    val result = sql(statement, session).result
    try {
      assert(result.size == rows.size)
      result.zip(rows).foreach { case (r1, r2) =>
        assert(r1.size == r2.size)
        r1.zip(r2).foreach { case (v1, v2) =>
          v1 shouldBe v2
        }
      }
    } catch {
      case _: AssertionError =>
        fail(s"expected:<$rows>, but was:<$result>")
      case e =>
        throw e
    }
  }
}

class SqlResult(private val statement: String, private val table: DBTable) {

  lazy val result: List[List[Any]] = {
    val rs = ListBuffer[List[Any]]()
    (0 until table.getRowCount).foreach { r =>
      rs += table.getColumns.asScala.map(_.getTypeValues.get(r)).toList
    }
    rs.toList
  }

  lazy val size: Int = this.result.size

  def show(): SqlResult = {
    table.show()
    this
  }

  override def toString: String = {
    this.result.toString()
  }

  override def hashCode(): Int = {
    this.result.hashCode()
  }

  override def equals(other: Any): Boolean = {
    this.result == other
  }
}

final case class SqlSession(var currentSchema: String = Database.DEFAULT_DATABASE)
